{
  Object buildId=idGenerator.generateId();
  List<DaemonInitialConnectException> accumulatedExceptions=Lists.newArrayList();
  int saneNumberOfAttempts=100;
  for (int i=1; i < saneNumberOfAttempts; i++) {
    final DaemonClientConnection connection=connector.connect(compatibilitySpec);
    try {
      Build build=new Build(buildId,connection.getDaemon().getToken(),action,requestContext.getClient(),requestContext.getBuildTimeClock().getStartTime(),parameters);
      return executeBuild(build,connection,requestContext.getCancellationToken(),requestContext.getEventConsumer());
    }
 catch (    DaemonInitialConnectException e) {
      LOGGER.debug("{}, Trying a different daemon...",e.getMessage());
      accumulatedExceptions.add(e);
    }
 finally {
      connection.stop();
    }
  }
  throw new NoUsableDaemonFoundException("Unable to find a usable idle daemon. I have connected to " + saneNumberOfAttempts + " different daemons but I could not use any of them to run the build. BuildActionParameters were "+ parameters+ ".",accumulatedExceptions);
}
